### PR TITLE
Add support for assignments in while-loops

### PR DESCRIPTION
**Pull request by ovdiiuv**
This PR adds support for assignments in while-loops. It also enables to combine multiple assignments in the while condition by adding support for some logical operators.

Fixes: #913

### CODE REVIEW

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 1]
```diff
@@ -1627,20 +1627,45 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
+    condVarRes = DifferentiateVarDecl(condVar, /*ignoreInit=*/true);
     condVarClone = condVarRes.getDecl();
+    if (condVarRes.getDecl_dx())
+      addToCurrentBlock(BuildDeclStmt(condVarRes.getDecl_dx()));
+    auto *condInit = condVarClone->getInit();
+    condVarClone->setInit(nullptr);
+    cond = BuildOp(BO_Assign, BuildDeclRef(condVarClone), condInit);
+    addToCurrentBlock(BuildDeclStmt(condVarClone));
+  }
+  // differentiate condition, allows assignments in the condition
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1646]:
```suggestion
  // Assignments in the condition are allowed, differentiate.
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 2]
```diff
@@ -1627,20 +1627,45 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
+  if (cond) {
+    cond = cond->IgnoreParenImpCasts();
+    auto* condBO = dyn_cast<BinaryOperator>(cond);
+    auto* condUO = dyn_cast<UnaryOperator>(cond);
+    // here is determined if condition differentiation is supported
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1651]:
```suggestion
    // FIXME: Currently we only support logical and assignment operators.
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 3]
```diff
@@ -1627,20 +1627,45 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
+    if ((condBO && (condBO->isLogicalOp() || condBO->isAssignmentOp())) ||
+        condUO) {
+      auto condDiff = Visit(cond);
+      // here is checked if condition is of the form
+      // ... <logicalOperator> ...
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1656]: Worth elaborating. Please write full sentences and maybe give a more complete example.

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 4]
```diff
@@ -1656,29 +1681,6 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
-  //   ...
-  // }
-  //
-  // gets differentiated to,
-  //
-  // while (double b = a) {
-  //   double _d_b = _d_a;
-  //   ...
-  //   ...
-  // }
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1676]: Still not sure why this comment is dropped rather than updated.

Comment (ovdiiuv) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1676]: well, it used to throw the  differentiation of the variable declaration in the body, but it doesn’t anymore. Now we do it before the while loop and I’ve added the documentation where it is done

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 5]
```diff
@@ -1627,20 +1627,45 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
   DeclDiff<VarDecl> condVarRes;
+
+  StmtDiff condDiff = Clone(WS->getCond());
+  Expr* cond = condDiff.getExpr();
+
+  // check if the condition contais a variable declaration and create a
+  // declaration of both the variable and it's adjoint before the while-loop
   if (condVar) {
-    condVarRes = DifferentiateVarDecl(condVar);
+    condVarRes = DifferentiateVarDecl(condVar, true);
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1637]: These do not seem resolved.

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 6]
```diff
@@ -1627,20 +1627,45 @@
...
     condVarClone = condVarRes.getDecl();
+    if (condVarRes.getDecl_dx())
+      addToCurrentBlock(BuildDeclStmt(condVarRes.getDecl_dx()));
+    auto condInit = condVarClone->getInit();
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1641]: Likewise.

File: include/clad/Differentiator/BaseForwardModeVisitor.h [hunk 7]
```diff
@@ -85,6 +85,7 @@ class BaseForwardModeVisitor
                          set to 0",
          args);
   }
+
```
Comment (vgvassilev) [inline include/clad/Differentiator/BaseForwardModeVisitor.h:88]:
```suggestion
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 8]
```diff
@@ -1627,20 +1627,53 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
   const VarDecl* condVar = WS->getConditionVariable();
   VarDecl* condVarClone = nullptr;
   DeclDiff<VarDecl> condVarRes;
+
+  StmtDiff condDiff = Clone(WS->getCond());
+  Expr* cond = condDiff.getExpr();
+
+  // check if the condition contais a variable declaration and create a
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1634]:
```suggestion
  // Check if the condition contais a variable declaration and create a
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 9]
```diff
@@ -1627,20 +1627,53 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
+  // declaration of both the variable and it's adjoint before the while-loop
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1635]:
```suggestion
  // declaration of both the variable and it's adjoint before the while-loop.
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 10]
```diff
@@ -1627,20 +1627,53 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
+  // Assignments in the condition are allowed, differentiate.
+  if (cond) {
+    cond = cond->IgnoreParenImpCasts();
+    auto* condBO = dyn_cast<BinaryOperator>(cond);
+    auto* condUO = dyn_cast<UnaryOperator>(cond);
+    // FIXME: Currently we only support logical and assignment operators.
+    if ((condBO && (condBO->isLogicalOp() || condBO->isAssignmentOp())) ||
+        condUO) {
+      auto condDiff = Visit(cond);
+      // after Visit(cond) is called the derivative could either be recorded in
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1655]:
```suggestion
      // After Visit(cond) is called the derivative could either be recorded in
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 11]
```diff
@@ -1627,20 +1627,53 @@ StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
...
+      // condDiff.getExpr() or condDiff.getExpr_dx(), hence we should build cond
+      // differently which is implemented below visiting statements like "(x=0)"
+      // records the differentiated statement in condDiff.getExpr_dx(), meaning
+      // we have to build in the form ((cond_dx), (cond)), wrapping cond_dx and
+      // cond into parentheses.
+      //
+      // visiting statements like "(x=0) || false" records the result in
+      // condDiff.getExpr(), meaning the differentiated condition is already
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1663]:
```suggestion
      // Visiting statements like "(x=0) || false" records the result in
      // condDiff.getExpr(), meaning the differentiated condition is already.
```

**Review by vgvassilev — APPROVED** [review]

**Pull request merged**